java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Search for the range of a particular element in the target vector.
+ */
+public class VectorRangeSearcher {
+
+  /**
+   * Search for the first occurrence of an element.
+   * The search is based on the binary search algorithm. So the target vector must be sorted.
+   * @param targetVector the vector from which to perform the search.
+   * @param comparator the criterion for the comparison.
+   * @param keyVector the vector containing the element to search.
+   * @param keyIndex the index of the search key in the key vector.
+   * @param <V> the vector type.
+   * @return the index of the first matched element if any, and -1 otherwise.
+   */
+  public static <V extends ValueVector> int getLowerBound(
+          V targetVector, VectorValueComparator<V> comparator, V keyVector, int keyIndex) {
+    comparator.attachVectors(keyVector, targetVector);
+
+    int low = 0;
+    int high = targetVector.getValueCount() - 1;
+
+    while (low <= high) {
+      int mid = (low + high) / 2;
+      int result = comparator.compare(keyIndex, mid);
+      if (result < 0) {
+        // the key is smaller
+        high = mid - 1;
+      } else if (result > 0) {
+        // the key is larger
+        low = mid + 1;
+      } else {
+        // the key equals the mid value, find the lower bound by going left-ward.
+
+        // compare with the left neighbour
+        int left = mid - 1;
+        if (left == -1) {
+          // this is the first value in the vector
+          return mid;
+        } else {
+          int leftResult = comparator.compare(keyIndex, left);
+          if (leftResult > 0) {
+            // the key is greater than the left neighbour, and equal to the current one
+            // we find it
+            return mid;
+          } else if (leftResult == 0) {
+            // the left neighbour is also equal, continue to go left
+            high = mid - 1;
+          } else {
+            // the key is larger than the left neighbour, this is not possible
+            throw new IllegalStateException("The target vector is not sorted ");
+          }
+        }
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * Search for the last occurrence of an element.
+   * The search is based on the binary search algorithm. So the target vector must be sorted.
+   * @param targetVector the vector from which to perform the search.
+   * @param comparator the criterion for the comparison.
+   * @param keyVector the vector containing the element to search.
+   * @param keyIndex the index of the search key in the key vector.
+   * @param <V> the vector type.
+   * @return the index of the last matched element if any, and -1 otherwise.
+   */
+  public static <V extends ValueVector> int getUpperBound(
+          V targetVector, VectorValueComparator<V> comparator, V keyVector, int keyIndex) {
+    comparator.attachVectors(keyVector, targetVector);
+
+    int low = 0;
+    int high = targetVector.getValueCount() - 1;
+
+    while (low <= high) {
```

Comment:
is it possible to factor out some duplicate code?

Reply:
It is possible, however, it may not be good for readability and code structure.

This is because, the only difference is in the case that the target element is equal to the source element. In this case, we either go left-ward or right-ward by binary search. 

So to extract the common code, we need to extract the logic for going left-ward or right-ward. This involves passing and returning multiple states: low, mid, high, plus a flag indicating is we have found the bound. This makes the code less clear.